src/objective-c/RxLibrary/GRXBufferedPipe.m

```diff
@@ -48,99 +52,109 @@ + (instancetype)pipe {
 
 - (instancetype)init {
   if (self = [super init]) {
-    _queue = [NSMutableArray array];
     _state = GRXWriterStateNotStarted;
+    _writeQueue = dispatch_queue_create(NULL, DISPATCH_QUEUE_SERIAL);
+    self.inputIsFinished = NO;
+    dispatch_suspend(_writeQueue);
   }
   return self;
 }
 
-- (id)popValue {
-  id value = _queue[0];
-  [_queue removeObjectAtIndex:0];
-  return value;
-}
-
-- (void)writeBufferUntilPausedOrStopped {
-  while (_state == GRXWriterStateStarted && _queue.count > 0) {
-    [_writeable writeValue:[self popValue]];
-  }
-  if (_inputIsFinished && _queue.count == 0) {
-    // Our writer finished normally while we were paused or not-started-yet.
-    [self finishWithError:_errorOrNil];
-  }
-}
-
 #pragma mark GRXWriteable implementation
 
-// Returns whether events can be simply propagated to the other end of the pipe.
-- (BOOL)shouldFastForward {
-  return _state == GRXWriterStateStarted && _queue.count == 0;
-}
-
 - (void)writeValue:(id)value {
-  if (self.shouldFastForward) {
-    // Skip the queue.
-    [_writeable writeValue:value];
-  } else {
+  if (self.inputIsFinished) {
+    return;
+  }
```

Comment:
1) It can
2) It will not be a problem because
 a) `writesFinishedWithError` may only be invoked once (assuming that is an invocation agreement) so `inputIsFinished`is not going to be set in another thread, and
 b) if `writeValue` won the race, the subsequent `writeValue` and `writeFinishedWithError` to `self.writeable` are still synchronized by `_writeQueue`.